### PR TITLE
Implement NativeLibrary.GetEntryPointModuleHandle

### LINKED ISSUE #56331: [API Proposal]: Provide an API to get the native module handle of the native process entry-point module
### Background and motivation

As mentioned in #7267, there is a strong desire for enabling using the "__Internal" name in a DllImport to point to methods in the hosting executable module of the process in both CoreCLR and Mono. Mono has had this support for many years, but CoreCLR doesn't have built-in support, and we don't have a desire to provide built-in support for this "hidden" name.

However, we understand the need for a way to resolve this name to the correct API. We propose adding an API that gets the native module handle of the entry-point module.

### API Proposal

```C#
namespace System.Runtime.InteropServices
{
     public static class NativeLibrary
     {
          public static IntPtr GetEntryPointModuleHandle();
     }
}
```     

Alternatively, we considered allowing `null` as a parameter to `NativeLibrary.Load(string)` to match how the implementation on non-Windows is `dlopen(NULL)`. The interop team decided that design was an abstraction leak and undesirable since the Windows implementation doesn't follow the same model. Additionally, the other `NativeLibrary.Load` overloads have additional behaviors and aren't thin wrappers, which makes the `null` option less desirable.

### API Usage

```C#
NativeLibrary.SetDllImportResolver(assembly, (name, _, paths) =>
{
     if (name == "__Internal")
     {
          return NativeLibrary.GetEntryPointModuleHandle();
     }
     return IntPtr.Zero;
});
```     

### Risks

None

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Fixes #56331
Fixes https://github.com/dotnet/runtime/issues/7267

I know there was a continuing discussion on naming, so we can continue that here if we so desire.

cc: @jkotas @AaronRobinsonMSFT @lambdageek

### CODE REVIEW

File: src/coreclr/hosts/corerun/corerun.cpp
Comment (AaronRobinsonMSFT) [inline src/coreclr/hosts/corerun/corerun.cpp:558]:
```suggestion
extern "C" DLL_EXPORT HRESULT __cdecl GetCurrentClrDetails(void** clrInstance, unsigned int* appDomainId)
```

File: src/coreclr/vm/ecalllist.h
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/ecalllist.h:801]:
```suggestion
	  QCFuncElement("GetEntryPointModuleHandleInternal", NativeLibraryNative::GetEntryPointModuleHandle)
```

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs:4]: Is this needed?

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs:275]: Is the error returned here really HResult?

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs:275]: I don't think so. It would typically be a Windows error code or an errno value. Which means it wouldn't normally have the error bit set.

File: src/tests/Interop/NativeLibrary/API/NativeLibraryTests.cs
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/NativeLibrary/API/NativeLibraryTests.cs:186]: I'd also like to see a negative test so we validate the exception code paths. Perhaps search for `"ExportThatDoesntExist"` too?

File: src/mono/mono/metadata/qcall-def.h
Comment (jkotas) [inline src/mono/mono/metadata/qcall-def.h:19]: Typo: NativeLibary

File: src/tests/Interop/NativeLibrary/API/NativeLibraryTests.cs
Comment (jkotas) [inline src/tests/Interop/NativeLibrary/API/NativeLibraryTests.cs:385]: Also validate whether or not this handles symbols in non-exe modules? I believe the current implementation will return symbols in non-exe modules on Unix.

**Comment by jkotas** [comment]
> Bring in some more QCall infrastructure that matches the CoreCLR

In this specific case, it would be even better to move the unmanaged part to src/libraries/Native (it is done that way in NativeAOT already) so the whole thing can be shared. I understand that it would be more work.

**Comment by jkoritzinsky** [comment]
> > Bring in some more QCall infrastructure that matches the CoreCLR
> 
> In this specific case, it would be even better to move the unmanaged part to src/libraries/Native (it is done that way in NativeAOT already) so the whole thing can be shared. I understand that it would be more work.

I didn't want to do that initially since Mono already has some special handling for `__Internal` that I wanted to piggy-back on and it felt weird to have most of `NativeLibrary` be implemented in the runtimes but have one method implemented in `src/libraries/Native`, but I can move it there if we want.

If I do move it, I'd like to leave the mono QCall infra in this PR (or worst case move it into another PR) to make it easier to add QCalls to Mono in the future.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs
Comment (vargaz) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs:263]: The icalls are usually put into the vm specific files.

Comment (jkoritzinsky) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs:263]: One of the things I wanted to do with this PR was start a process of sharing QCalls between the runtimes as per #57369, Mono may be moving to more QCalls in the future. If CoreCLR and Mono are going to have the same signatures with the same behavior, we might as well share the managed implementation even if we can't share the native.

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs:263]: > #57369

I think this is wrong issue number.

FWIW, we have tried to use shared QCalls a few times in the past, but it always fired back for various reasons and got undone so far. We can give it another try though.

The solution that is proven to work well is to the cleanup the unmanaged dependencies and move them to libraries/src/native .

Comment (jkoritzinsky) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs:263]: Yes that is. I meant #57639, sorry.

I'll look at moving the dependencies to the libraries native sources.

Comment (jkoritzinsky) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs:263]: IIRC the last time QCalls were shared was for Globalization, which was undone due to the changes needed for single file providing an alternative solution.

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs:263]: There was also a brief period where QCalls used for tracing were shared.

File: src/libraries/Native/Unix/System.Native/pal_dl.c
Comment (stephentoub) [inline src/libraries/Native/Unix/System.Native/pal_dl.c:19]: I assume dlopen properly handles concurrent usage and is idempotent?

Comment (jkoritzinsky) [reply src/libraries/Native/Unix/System.Native/pal_dl.c:19]: Mono implements this the same way so I believe it does for this case.

Comment (jkotas) [reply src/libraries/Native/Unix/System.Native/pal_dl.c:19]: I would be useful to double check this assumption.

Comment (jkoritzinsky) [reply src/libraries/Native/Unix/System.Native/pal_dl.c:19]: I've done some extended testing locally and `dlopen(NULL, RTLD_LAZY)` returns the same value every time since it's the pseudohandle.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs:276]: This still does not look right for non-Windows. This assumes that the Unix native PInvoke errors codes are the same as Win32 error codes which they certainly are not.

**Comment by jkotas** [comment]
>  there was a continuing discussion on naming, so we can continue that here if we so desire.

Should we close on the naming? (I have shared my opinion in the issue - I do not have much to add.)

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs
Comment (jkoritzinsky) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs:276]: Do we have a generalized mechanism today for converting a Unix `errno` value to an exception or a pattern I should follow for converting the error code to an exception on non-Windows platforms?

**Comment by jkoritzinsky** [comment]
Since the [manpage for `dlopen`](https://linux.die.net/man/3/dlopen) uses the term "handle for the main program" for what's returned when you pass `NULL` to it, I think I'd be okay switching the name to `GetMainProgramLibraryHandle` or `GetMainProgramHandle`.

I'll tag in @bartonjs and @lambdageek for which of those two we prefer and so we can get API board approval for the new name.

**Comment by AaronRobinsonMSFT** [comment]
My vote is for `GetMainProgramHandle`.

**Comment by jkoritzinsky** [comment]
I'll send an email to fxdc to see if we can get the name changed.

**Comment by bartonjs** [comment]
Personally, I don't have strong feelings between `GetEntryPointModuleHandle`, `GetMainProgramHandle`, and `GetMainProgramLibraryHandle`.  I'll even square the triangle with `GetEntryPointLibraryHandle`.

I do have weak feelings, which produce a preference order.

* GetEntryPointModuleHandle: The entrypoint is a unique thing.  You can have multiple methods named Main, but only one is the actual entrypoint.
* GetEntryPointLibraryHandle: While I think of .netmodules before I think of any Windows-specific use of "Module", I'm happy to say "library" instead of "module" here.  Maybe I'm wrong and this is my top choice.
* GetMainProgramLibraryHandle: In addition to Main being a non-unique target, lowercase "main" is subjective.  If you wrote a thin exe that hosts a big worker library, you may think of the worker library as the "main" aspect of your program.  It's probably not a likely mistake, but I like precision.
* GetMainProgramHandle: This could be a method handle, a function pointer handle, a library handle, a process handle, ... anything.  I don't like how vague it is.  Though, given that the type is `NativeLibrary`, maybe it's OK, and maybe it's better than the longer version.

Now that I'm done, I think GetEntryPointLibraryHandle is, in fact, my probably weak top choice (it's, at most, "barely strong").  But I don't want to rewrite my narrative bullet points.

**Comment by AaronRobinsonMSFT** [comment]
I have no problem with `GetEntryPointLibraryHandle`. It is slightly misleading since on non-Windows platforms it is a pseudo handle, but if we are being honest users of this API are going to be people who are trying to do something very specific and likely know what is going on under the covers.

I still prefer `GetMainProgramHandle` since "Program" is vague enough it can accommodate the the pseudo handle idea.

**Comment by jkotas** [comment]
@bartonjs The handle returned by this method on Unix is pseudo handle that corresponds to set of files (multiple libraries) on disk: It is the main program binary and all .so/.dylib files that the main program binary links to.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs
Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs:276]: > Do we have a generalized mechanism today for converting a Unix errno value to an exception

We do not. (The closest one is `GetExceptionForIoErrno`, but it is very IO scenario focused and it may not produce the right result here.)

I think we typically use `throw new Win32Exception()` in situation like this one that are expected to never throw in practice. Despite its name, Win32Exception on non-Windows treats the error code as the platform specific error code (not as Windows error code).

File: src/libraries/Native/Unix/System.Native/pal_dl.c
Comment (am11) [reply src/libraries/Native/Unix/System.Native/pal_dl.c:19]: There is a slight cost per the [`dlopen(3)`](https://linux.die.net/man/3/dlopen) manpage:

> If the same library is loaded again with dlopen(), the same file handle is returned. The dl library maintains reference counts for library handles, so a dynamic library is not deallocated until dlclose() has been called on it as many times as dlopen() has succeeded on it. The _init() routine, if present, is only called once. But a subsequent call with RTLD_NOW may force symbol resolution for a library earlier loaded with RTLD_LAZY. 

so it is not completely free if we don't call `dlclose()` as many times (where possible; as in some cases we don't).

Comment (AaronRobinsonMSFT) [reply src/libraries/Native/Unix/System.Native/pal_dl.c:19]: I'm not sure that applies to the case where `NULL` is passed in. This isn't for a library but for the main program. The few implementations of `dlopen` I've looked at appear to special case `NULL` and return very early.

Comment (am11) [reply src/libraries/Native/Unix/System.Native/pal_dl.c:19]: Interesting that manpage does not specify the special leeway for `dlopen(NULL,..)` that we are free to skip `dlopen()`.
I only looked at glibc sources and it seems to at least increment/decrement reference counts (`l_direct_opencount`) regardless of `NULL`, as in; `dlclose()` on a handle returned by `dlopen(NULL)` is not a no-op.

Comment (AaronRobinsonMSFT) [reply src/libraries/Native/Unix/System.Native/pal_dl.c:19]: @am11 Agreed. I just confirmed your observation in glibc as well. Can't really argue with the fact that it does appear to increment and decrement, at least in the most common implementation. I did note the comment on the dummy for the executable, `_dl_main_map`, seems to indicate this is a very minimal instance. Perhaps the inc/dec doesn't impact this in the same way?

Comment (AaronRobinsonMSFT) [reply src/libraries/Native/Unix/System.Native/pal_dl.c:19]: I'm thinking `lt_executable` is the sentinel here.

**Comment by jkoritzinsky** [comment]
This is blocked on #58529

**Comment by jkoritzinsky** [comment]
It looks like macOS doesn't seem to support the global symbol lookup. @lambdageek any ideas why this lookup doesn't work on Mac? (The docs say that it should)

**Comment by AaronRobinsonMSFT** [comment]
I'm converting this issue to Draft.

**Comment by jkoritzinsky** [comment]
All of the test failures are unrelated. This is finally ready for review!

File: src/tests/Interop/NativeLibrary/NativeLibraryToLoad/GlobalLoadHelper.cpp
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/NativeLibrary/NativeLibraryToLoad/GlobalLoadHelper.cpp:6]: Perhaps add this to `platformdefines.h`?

File: src/tests/Interop/NativeLibrary/NativeLibraryToLoad/CMakeLists.txt
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/NativeLibrary/NativeLibraryToLoad/CMakeLists.txt:3]: ?

File: src/tests/Interop/NativeLibrary/NativeLibraryToLoad/GlobalLoadHelper.cpp
Comment (jkoritzinsky) [reply src/tests/Interop/NativeLibrary/NativeLibraryToLoad/GlobalLoadHelper.cpp:6]: I don't think we should add this to `platformdefines.h` as we generally use that more as a PAL layer and we wouldn't be providing cross-plat wrappers around these methods.

File: src/tests/Interop/NativeLibrary/API/GetMainProgramHandleTests.cs
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/NativeLibrary/API/GetMainProgramHandleTests.cs:12]: I'd like to see a test that validates passing the result from `GetMainProgramHandle()` to `Free()`. We should then also document the behavior unless it is a nop - which is what I'd assume it to be.

**Review by AaronRobinsonMSFT — APPROVED** [review]

File: src/native/libs/System.Native/pal_dynamicload.c
Comment (jkotas) [inline src/native/libs/System.Native/pal_dynamicload.c:63]: This has potential race condition depending on how the C++ decides to compile it. The C/C++ compiler is free to reorder the if check and reading of `g_defaultSearchOrderPseudoHandle` to return. It means that the method can return NULL instead of the actual value.

This is real problem. We had many bugs like that over the years - they are always very hard to trace down.

Comment (jkoritzinsky) [reply src/native/libs/System.Native/pal_dynamicload.c:63]: This can be fixed by making `g_defaultSearchOrderPseudoHandle` volatile, correct?

Comment (jkotas) [reply src/native/libs/System.Native/pal_dynamicload.c:63]: Yup. Also, you may want to cache the value in a local so that it does not need to be fetched from memory twice.

**Pull request merged**
